src/peft/tuners/lora/layer.py

```diff
@@ -434,7 +446,14 @@ def forward(self, x: torch.Tensor, *args: Any, **kwargs: Any) -> torch.Tensor:
                 x = x.to(lora_A.weight.dtype)
 
                 if not self.use_dora[active_adapter]:
-                    result = result + lora_B(lora_A(dropout(x))) * scaling
+                    if _xlora_layer is not None:
```

Comment:
I'm not so happy with this addition to the `LoraLayer`s. It makes reading and understanding them more complex and requires all LoRA layers to be updated (e.g. what about the bnb lora layers?).

I couldn't come up with an alternative yet, but I wonder if we could achieve something with wrapping and/or forward hooks. I'll continue to think about this tomorrow but wanted to let you know already in case you have some ideas.

Reply:
I agree, it is not very easy to read. Perhaps we could implement some sort of hook in the `LoraLayer` so that techniques such as DoRA and X-LoRA could use that instead of modifying the layer source?